Saving points of interest

i-zor066 opened the pull request:
This is where I'm currently at.

i-zor066 committed:
Added DataBaseOpenHelper, Table and PointsOfInterestTable.

i-zor066 committed:
Added DataSource.

i-zor066 committed:
Added data recovery methods and a Log.v "test" (writing to, and recovering stuff from the database works.

i-zor066 committed:
Added BlocSpotApplication, PointsOfInterest class and Tests.

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/BlocSpotApplication.java:
```diff
@@ -0,0 +1,33 @@
+        return BlocSpotApplication.getSharedInstance().getDataSource();
+    }
+
+    private static BlocSpotApplication sharedInstance;
```
Fields should be at top of the class, so in general the order within a class is:
1. static fields (class fields)
2. instance fields
3. methods

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/DataSource.java:
```diff
@@ -0,0 +1,52 @@
+    private DatabaseOpenHelper databaseOpenHelper;
+    private PointsOfInterestTable pointsOfInterestTable;
+
+
```
The extra space here (lines 20 and 21) is unnecessary here. Don't forget to format the code from time to time (`⌘+⌥+L` - with the normal Android Studio key bindings).

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/DataSource.java:
```diff
@@ -0,0 +1,52 @@
+        pointsOfInterestTable = new PointsOfInterestTable();
+        databaseOpenHelper = new DatabaseOpenHelper(context, pointsOfInterestTable);
+
+        if (BuildConfig.DEBUG && true) {
```
As `BuildConfig.DEBUG` already returns a boolean, there is no need of asserting with an additional `&& true`. 
`if (BuildConfig.DEBUG)` alone is sufficient.

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/DataSource.java:
```diff
@@ -0,0 +1,52 @@
+        databaseOpenHelper = new DatabaseOpenHelper(context, pointsOfInterestTable);
+
+        if (BuildConfig.DEBUG && true) {
+            context.deleteDatabase("blocspot_db");
```
It is a good practise to extract this as a constant in order to have only one point of edit (for example if you decide to change the name of your database, you should only change it once in one place).
I see now that actually it is extracted [here](https://github.com/i-zor066/BlocSpot/pull/1/files#diff-7b610108acc9fc52c7141e4a1a7a16f5R14). As a first step you could declare that variable `public` and reference it here as `context.deleteDatabase(DatabaseOpenHelper.NAME)`. This would eliminate the possibility to change the name in one place, but to forget to change it in the other one.

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/DataSource.java:
```diff
@@ -0,0 +1,52 @@
+            SQLiteDatabase writableDatabase = databaseOpenHelper.getWritableDatabase();
+            new PointsOfInterestTable.Builder()
+                    .setTitle("")
+                    .insert(writableDatabase);
```
Why would you insert an empty entry in the database when the `DateSource` is created?

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/model/PointsOfInterest.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Created by igor on 22/9/15.
+ */
+public class PointsOfInterest {
```
I would incline to name this `PointOfInterest` rather then Point**s**. The plural often means that the class represents a collection of items, but in this case it doesn't, it's just one point of interest.

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/model/PointsOfInterest.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class PointsOfInterest {
+
+    private String title;
```
All these fields should be `final` in order to promote immutability. A good read on why immutability is good can be found [here](http://www.yegor256.com/2014/06/09/objects-should-be-immutable.html).

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/DataSource.java:
```diff
@@ -0,0 +1,52 @@
+        Cursor cursor = PointsOfInterestTable.getRowFromTitle(databaseOpenHelper.getReadableDatabase(), title);
+        cursor.moveToFirst();
+        PointsOfInterest pointsOfInterest = new PointsOfInterest(title, PointsOfInterestTable.getAddress(cursor), PointsOfInterestTable.getLatitude(cursor), PointsOfInterestTable.getLongitude(cursor));
+        return pointsOfInterest;
```
:exclamation: After usage (i.e. after creating the `pointsOfInterest` object) the cursor must be closed - `cursor.close()` so that we don't clutter the database once we don't need that particular cursor.

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/DataSource.java:
```diff
@@ -0,0 +1,52 @@
+    public PointsOfInterest getPOIfromDBwithTitle (String title) {
+        Cursor cursor = PointsOfInterestTable.getRowFromTitle(databaseOpenHelper.getReadableDatabase(), title);
+        cursor.moveToFirst();
+        PointsOfInterest pointsOfInterest = new PointsOfInterest(title, PointsOfInterestTable.getAddress(cursor), PointsOfInterestTable.getLatitude(cursor), PointsOfInterestTable.getLongitude(cursor));
```
This would read much better if the other parameters apart from the `title` would be extracted as well. Something like:

``` java
String address = PointsOfInterestTable.getAddress(cursor);
float latitude = PointsOfInterestTable.getLatitude(cursor);
float longitude = PointsOfInterestTable.getLongitude(cursor);
PointsOfInterest pointsOfInterest = new PointsOfInterest(title, address, latitude, longitude);
```

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/model/database/table/PointsOfInterestTable.java:
```diff
@@ -0,0 +1,91 @@
+                new String[]{title}, null, null, null, null);
+    }
+
+    private static final String NAME ="points_of_interest";
```
As I mentioned before, these fields (which are in fact constants here), should be moved at the top of the class.

Electryc commented on BlocSpot/app/src/main/java/com/izor066/android/blocspot/api/model/database/table/Table.java:
```diff
@@ -0,0 +1,63 @@
+
+    public abstract String getCreateStatement();
+
+    public void onUpgrade(SQLiteDatabase writableDatabase, int oldVersion, int newVersion) {
```
Do you plan to implement this method? If yes, a comment with a `//TODO implement upgrade path` (or something like this) would flag that this method is left un-implemented (yet) on purpose. Otherwise, if you're not sure, simply delete it until you need it. Leaving unimplemented methods in your code, especially when working within a team, can create a bit of confusion. 

[YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) is a good principle to follow in software development.

Electryc commented:
A good practice to follow when submitting pull requests (especially when working in a team) is to provide a detailed description what the PR is about. In this particular case, it could have been something like:

> This PR focuses on persisting points of interest into the database. 
> In order to do this, the following major entities have been implemented:
> - `DatabaseOpenHelper`: responsible with creating the necessary tables
> - `Table` abstraction with its `PointsOfInterest` implementation
> - model for `PointsOfInterest` entry 
> - `DataSource`: responsible with inserting and retrieving `PointsOfInterest` to/from the database
>   Additionally, some tests were added.

Of course, you don't have always to mention all of the classes that you've implemented in the PR, but only maybe the major ones, just to reiterate over their responsibilities. 

All of the above are very helpful when working in a team: 
- firstly when someone is reviewing your PR, he has a better understanding of what you've tried to achieve
- secondly the PRs are like a history of the application implementation - at any point I can go back and re-read a PR, or if I'm not sure why a commit was done, I can get to the PR of that particular commit and just by reading the description I understand the context. 

:+1: Most of my comments on this PR are not critical, the only one that makes me not to merge it is the closing cursor one. If you could address all of the comments it would be great, but the critical one is the closing cursor. After that one is addressed, I would merge this so that we keep the implementation flow going.
Don't forget that you can always branch out of this branch in order to advance with the implementation of the following assignment, so you don't have to wait until this is merged in.

i-zor066 committed:
Implemented most of the comments:
- Closed the cursor where required.
- put fields to the top of the class in BlocSpotApplication.java and PointsOfInterestTable
- Removed extra lines in DataSource.java
- Removed (&& true) from DataSource.java
- Removed empty entry insertion from DataSource.java
- Changed the name of PointsOfInterest.java to PointOfInterest.java
- Changed fields to final in PointOfInterest.java
- Extracted parameters in getPOIfromDBwithTitle

Tested, nothing broke.

Electryc commented:
Normally, when there are comments on a PR, after addressing them it's nice to put a comment that "the comments have been addressed" or something like that and the reviewer can take a look again and decide to merge or not based on the latest fixes. So no self merges :tongue:

Electryc commented:
:+1: But it's alright, don't worry.
